nemo_curator/modules/semdedup/compute_embeddings.py

```diff
@@ -0,0 +1,170 @@
+# limitations under the License.
+
+import os
+import time
+from dataclasses import dataclass
+
+import torch
+import torch.nn as nn
+from crossfit import op
+from crossfit.backend.torch.hf.model import HFModel
+from torch.nn import functional as F
+from transformers import AutoConfig, AutoModel, AutoTokenizer
+
+from nemo_curator.modules.semdedup.utils import parse_arguments
+from nemo_curator.utils.distributed_utils import get_client, read_data, write_to_disk
+from nemo_curator.utils.file_utils import get_remaining_files
+from nemo_curator.utils.script_utils import parse_client_args
+
+
+def mean_pooling(model_output, attention_mask):
+    token_embeddings = model_output[0]
+    input_mask_expanded = (
+        attention_mask.unsqueeze(-1).expand(token_embeddings.size()).float()
+    )
+
+    sum_embeddings = torch.sum(token_embeddings * input_mask_expanded, dim=1)
+    sum_mask = torch.clamp(input_mask_expanded.sum(dim=1), min=1e-9)
+
+    return sum_embeddings / sum_mask
+
+
+## Define the model config
+@dataclass
+class EmbeddingConfig:
+    path_or_name: str
+    max_mem_gb: int
+    max_seq_length: int = None
+
+    def __post_init__(self):
+        # Set max_seq_length based on model's capabilities
+        self.max_seq_length = AutoTokenizer.from_pretrained(
+            self.path_or_name
+        ).model_max_length
+        # Guard against excessively large max lengths
+        if self.max_seq_length > 1e5:
+            self.max_seq_length = AutoConfig.from_pretrained(
+                self.path_or_name
+            ).max_position_embeddings
+
+
+class CustomModel(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        self.config = config
+        self.model = AutoModel.from_pretrained(config.path_or_name, config=self.config)
+
+    def feature(self, input_ids, attention_mask):
+        with torch.autocast(device_type=input_ids.device.type):
+            embeddings = self.model(input_ids=input_ids, attention_mask=attention_mask)
+        return embeddings
+
+    @torch.no_grad()
+    def forward(self, batch):
+        feature = self.feature(batch["input_ids"], batch["attention_mask"])
+        emb = mean_pooling(feature, batch["attention_mask"])
+        return F.normalize(emb, dim=1)
+
+
+class CrossFitModel(HFModel):
+    def __init__(self, config: EmbeddingConfig):
+        self.config = config
+        super().__init__(self.config.path_or_name, max_mem_gb=self.config.max_mem_gb)
+
+    def load_model(self, device="cuda"):
+        model = CustomModel(self.config)
+        model = model.to(device)
+        model.eval()
+        return model
+
+    def load_config(self):
+        return AutoConfig.from_pretrained(self.path_or_name)
+
+    def max_seq_length(self):
+        return self.config.max_seq_length
+
+
+def main():
+    args = parse_arguments()
+    sample = args.sample
+    args_emb = args.embeddings
+
+    st = time.time()
+
+    input_data_dir = args_emb["input_data_dir"]
```

Comment:
input_data_dir should be set to os.path.join(args.root, args_emb["input_data_dir"])